tests/templates/test_subroutines/test_trotter.py

```diff
@@ -468,3 +511,330 @@ def circ():
+    def test_torch_gradient(self, order, n):
+        """Test that the gradient is computed correctly using torch"""
+        import torch
+
+        time = torch.tensor(1.5, dtype=torch.complex64, requires_grad=True)
+        coeffs = torch.tensor([1.23, -0.45], dtype=torch.complex64, requires_grad=True)
+        time_reference = torch.tensor(1.5, dtype=torch.complex64, requires_grad=True)
+        coeffs_reference = torch.tensor([1.23, -0.45], dtype=torch.complex64, requires_grad=True)
+        terms = [qml.PauliX(0), qml.PauliZ(0)]
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circ(time, coeffs):
+            h = qml.dot(coeffs, terms)
+            qml.TrotterProduct(h, time, n=n, order=order)
+            return qml.expval(qml.Hadamard(0))
+
+        @qml.qnode(dev)
+        def reference_circ(time, coeffs):
+            with qml.QueuingManager.stop_recording():
+                decomp = _generate_simple_decomp(coeffs, terms, time, order, n)
+
+            for op in decomp[-1::-1]:
+                qml.apply(op)
+
+            return qml.expval(qml.Hadamard(0))
+
+        res_circ = circ(time, coeffs)
+        res_circ.backward()
+        measured_time_grad = time.grad
+        measured_coeff_grad = coeffs.grad
+
+        ref_circ = reference_circ(time_reference, coeffs_reference)
+        ref_circ.backward()
+        reference_time_grad = time_reference.grad
+        reference_coeff_grad = coeffs_reference.grad
+
+        assert allclose(measured_time_grad, reference_time_grad)
+        assert allclose(measured_coeff_grad, reference_coeff_grad)
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize("order, n", ((1, 1), (1, 2), (2, 1)))
+    def test_tf_gradient(self, order, n):
+        """Test that the gradient is computed correctly using tensorflow"""
+        import tensorflow as tf
+
+        time = tf.Variable(1.5, dtype=tf.complex128)
+        coeffs = tf.Variable([1.23, -0.45], dtype=tf.complex128)
+        terms = [qml.PauliX(0), qml.PauliZ(0)]
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circ(time, coeffs):
+            h = qml.sum(
+                qml.s_prod(coeffs[0], terms[0]),
+                qml.s_prod(coeffs[1], terms[1]),
+            )
+            qml.TrotterProduct(h, time, n=n, order=order)
+            return qml.expval(qml.Hadamard(0))
+
+        @qml.qnode(dev)
+        def reference_circ(time, coeffs):
+            with qml.QueuingManager.stop_recording():
+                decomp = _generate_simple_decomp(coeffs, terms, time, order, n)
+
+            for op in decomp[-1::-1]:
+                qml.apply(op)
+
+            return qml.expval(qml.Hadamard(0))
+
+        with tf.GradientTape() as tape:
+            result = circ(time, coeffs)
+
+        measured_time_grad, measured_coeff_grad = tape.gradient(result, (time, coeffs))
+
+        with tf.GradientTape() as tape:
+            result = reference_circ(time, coeffs)
```

Comment:
Is this the standard way to compute the gradients?

Reply:
For TensorFlow yes